core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1560,12 +1571,43 @@ public void run() {
 
     @Override
     public ManagedChannelBuilder<?> createResolvingOobChannelBuilder(String target) {
+      return createBuilder(target, null);
+    }
+
+    @Override
+    public ManagedChannelBuilder<?> createResolvingOobChannelBuilder(
+        String target, ChannelCredentials creds) {
+      return createBuilder(target, creds);
+    }
+
+    private ManagedChannelBuilder<?> createBuilder(
+        String target, @Nullable ChannelCredentials creds) {
       final class ResolvingOobChannelBuilder
           extends ForwardingChannelBuilder<ResolvingOobChannelBuilder> {
-        private final ManagedChannelImplBuilder managedChannelImplBuilder;
-
-        ResolvingOobChannelBuilder(String target) {
-          managedChannelImplBuilder = new ManagedChannelImplBuilder(target,
+        @Nullable final ManagedChannelImplBuilder managedChannelImplBuilder;
+        final ManagedChannelBuilder<?> delegate;
+        ClientTransportFactory transportFactory;
+
+        ResolvingOobChannelBuilder(String target, @Nullable ChannelCredentials channelCreds) {
+          CallCredentials callCredentials;
+          if (channelCreds == null) {
+            transportFactory = originalTransportFactory;
+            callCredentials = null;
+          } else {
+            SwapChannelCredentialsResult swapResult =
+                originalTransportFactory.swapChannelCredentials(channelCreds);
```

Comment:
You can probably simply the flow a by having `channelCreds == null` case produce a `SwapChannelCredentialsResult`.

```suggestion
          SwapChannelCredentialsResult swapResult;
          if (channelCreds == null) {
            swapResult = new SwapChannelCredentialsResult(originalTransportFactory, null);
          } else {
            swapResult = originalTransportFactory.swapChannelCredentials(channelCreds);
          }
```

Reply:
Well, although this does simplify the flow a little bit, but I would rather not to instantiate a `SwapChannelCredentialsResult` if it is not really a result of `ClientTransportFactory.swapChannelCredentials()`.